applications/minotari_merge_mining_proxy/src/block_template_protocol.rs

```diff
@@ -122,87 +126,56 @@ impl BlockTemplateProtocol<'_> {
-                                new_template.clone(),
-                                template_with_coinbase.clone(),
-                            )
-                            .await;
-
-                        (new_template, template_with_coinbase, height)
-                    },
-                    Some((new_template, template_with_coinbase)) => {
-                        let height = new_template
-                            .template
-                            .header
-                            .as_ref()
-                            .map(|h| h.height)
-                            .unwrap_or_default();
-                        debug!(target: LOG_TARGET, "Used existing new block template at height: #{} (try {})", height, loop_count);
-                        (new_template, template_with_coinbase, height)
+                        }
                     },
                 };
 
-                let block = match self.get_new_block(block_template_with_coinbase).await {
-                    Ok(b) => {
-                        debug!(
-                            target: LOG_TARGET,
-                            "Requested new block at height: #{} (try {}), block hash: `{}`",
-                            height, loop_count,
-                            {
-                                let block_header = b.block.as_ref().map(|b| b.header.as_ref()).unwrap_or_default();
-                                block_header.map(|h| h.hash.clone()).unwrap_or_default().to_hex()
-                            },
-                        );
-                        b
-                    },
-                    Err(MmProxyError::FailedPreconditionBlockLostRetry) => {
-                        debug!(
-                            target: LOG_TARGET,
-                            "Chain tip has progressed past template height {}. Fetching a new block template (try {}).",
-                            height, loop_count
-                        );
-                        continue;
-                    },
-                    Err(err) => {
-                        error!(target: LOG_TARGET, "grpc get_new_block ({})", err.to_string());
-                        return Err(err);
-                    },
-                };
+                let height = block
+                    .block
+                    .as_ref()
+                    .map(|b| b.header.as_ref().map(|h| h.height).unwrap_or_default())
+                    .unwrap_or_default();
 
-                (
-                    add_monero_data(block, monero_mining_data.clone(), new_template)?,
-                    height,
-                )
+                let miner_data = block
+                    .miner_data
+                    .as_ref()
+                    .cloned()
+                    .ok_or_else(|| MmProxyError::GrpcResponseMissingField("miner_data"))?;
+
+                (add_monero_data(block, monero_mining_data.clone(), miner_data)?, height)
```

Comment:
Should `new_template` as calculated earlier not be used for non-pools? I am not sure if miner data pulled form the block will be the same. In `fn add_monero_data(` parameter `template_data: NewBlockTemplateData` is fed in and not pulled from `tari_block_result: grpc::GetNewBlockResult`. With the suggestion below we know the code will behave the same.
```suggestion
                match self.p2pool_client.as_mut() {
                    Some(p2pool_client) => {
                        (add_monero_data(block, monero_mining_data.clone(), miner_data)?, height)
                    },
                    None => {
                        (add_monero_data(block, monero_mining_data.clone(), new_template)?, height)
                    },
                }
```

Reply:
I'm not sure I understand what you mean here. I think the mmproxy caching can be simplified even for non pools, but I unfortunately don't have the time to do this now